terrain/block_thread_manager.h

```diff
@@ -466,12 +467,14 @@ class VoxelBlockThreadManager {
 				}
 			}
 
-			if (data.thread_exit) {
+			if (data.thread_exit && data.semaphore->get() >= 1) {
 				break;
 			}
 
 			// Wait for future wake-up
 			data.semaphore->wait();
+		
+			keep_going = !data.thread_exit || data.semaphore->get() > 0;
```

Comment:
I'm not sure of this. The reason why the semaphore is raised is to make the call to `wait()` resume the thread, so it will always be true here I think?
I believe the exit condition doesn't need to be driven by the semaphore.

Reply:
Lets assume normally when we enter thread, semaphore has value of 1. All work is done, and we reach line 475. After this line runs, semaphore will have value of 0, right?

However, I noticed the above is not always the case, frequently when autosaving, for some reason semaphore value is 2, and every time this happens, I noticed that my changes weren't saved.  Thus, the reason for `keep_going = !data.thread_exit || data.semaphore->get() > 0;`.

What honestly confuses me more, is that from further testing, it seems the following change is not needed at all: `if (data.thread_exit && data.semaphore->get() >= 1)`.

I'll have the updated PR shortly.

Reply:
I wonder if it doesnt just need `while(true)` on the outer loop instead of `while(!thread_exit)`, with a `break` if `thread_exit == true` in the place you added it.

Reply:
It seems you are correct :laughing: 
The only change needed is `while(!data.thread_exit) -> while(true)`.
I'll have that update in just a bit. 

Can you say why this is likely the case? I mean, how can the value of tread_exit change inside the loop?

Reply:
It can change because the main thread can set it anytime. It might be false at the beginning and become true while the loop is processing. That means while that processing happens, some new inputs could have arrived as well, so at the end the loop needs to restart one last time to be sure to catch what remains.

Reply:
Oh I see. Then I should move the break statement to be last in the loop, even after the semaphore wait call, right?

Reply:
Yeah, cuz it doesn't give the last shot as I described earlier.